pkg/agent/openflow/pipeline.go

```diff
@@ -3130,3 +2967,83 @@ func (c *featurePodConnectivity) l3FwdFlowToExternal(category cookie.Category) b
 		Action().GotoStage(binding.SwitchingStage).
 		Done()
 }
+
+// Feature: Service
+// Stage: RoutingStage
+// Tables: L3ForwardingTable
+// New added
+// l3FwdFlowToExternal generates the default L3 forward flows to forward packets of Service connections whose destination
+// is external network.
+func (c *featureService) l3FwdFlowToExternal(category cookie.Category) binding.Flow {
+	return L3ForwardingTable.ofTable.BuildFlow(priorityLow).
+		Cookie(c.cookieAllocator.Request(category).Raw()).
+		MatchRegMark(RewriteMACRegMark).
+		MatchCTMark(ServiceCTMark).
+		MatchRegMark(NotAntreaFlexibleIPAMRegMark).
+		Action().SetDstMAC(c.gatewayMAC).
+		Action().LoadRegMark(ToExternalRegMark).
+		Action().NextTable().
+		Done()
+}
+
+// Feature: Service
+// Stage: RoutingStage
+// Tables: ServiceHairpinMarkTable
+// New added
+// intraClusterHairpinFlow generates the flows to match request packets of hairpin Service connections inside the Cluster.
+func (c *featureService) intraClusterHairpinFlow(category cookie.Category, endpoint net.IP) binding.Flow {
+	ipProtocol := getIPProtocol(endpoint)
+	return ServiceHairpinMarkTable.ofTable.BuildFlow(priorityNormal).
+		Cookie(c.cookieAllocator.Request(category).Raw()).
+		MatchProtocol(ipProtocol).
+		MatchSrcIP(endpoint).
+		MatchDstIP(endpoint).
+		Action().LoadRegMark(HairpinRegMark).
+		Action().LoadRegMark(SNATWithGatewayIP).
+		Action().LoadRegMark(NotRequireSNATRegMark).
+		Action().NextTable().
+		Done()
+}
+
+// Feature: Service
+// Stage: RoutingStage
+// Tables: ServiceHairpinMarkTable
+// New added
+// externalClusterHairpinFlows generate the flow to match packets of hairpin Service connections outside the Cluster.
```

Comment:
s/flow/flows

Add more more comments to explain what kind of service connections needs hairpin

Reply:
Updated.